### PR TITLE
#0: Add graph-based query apis

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
None

### Problem description
Some of the C++ Test utils may be helpful outside of tests.

### What's changed
* Moved util functions to ttnn/graph_utils.hpp
* Addied new set of query apis in graph_operation_query.hpp
* Tests are extended

### Checklist
- [ ] Post commit CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/10484188200)

### CODE REVIEW

File: tests/ttnn/unit_tests/gtests/test_graph_add.cpp [hunk 1]
```diff
@@ -158,20 +50,35 @@ TEST_P(AddOpGraphTestFixture, AddGraphTrace) {
...
+
+        EXPECT_EQ(graph::extract_output_L1_size(json_trace), params.expected_output_L1_size);
+
+        // Query calls
+        auto query_peak_memory_load = graph::query_peak_memory_load([&] {
+            const auto output_tensor = ttnn::add(input_tensor_a, input_tensor_b);
+            return output_tensor;
+        });
+
+        auto query_output_L1_size = graph::query_output_L1_size([&] {
```
Comment (dmakoviichuk-tt) [inline tests/ttnn/unit_tests/gtests/test_graph_add.cpp:76]: maybe better have one query function which returns struct with L1, DRAM and etc

**Review by dmakoviichuk-tt — APPROVED** [review]
LGTM but porbably you can switch to one query function

File: ttnn/cpp/ttnn/graph/graph_trace_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+#include <unordered_set>
+#include <string>
+#include <cstdlib> // std::strtoul
+
+
+namespace ttnn::graph {
+
+namespace {
+ttnn::Shape parse_shape(const std::string_view& shape_string) {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/graph/graph_trace_utils.cpp:19]: you can pass string_view buy value.

Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/graph/graph_trace_utils.cpp:19]: Overall need to check if there is serialize/deserialize method for shape in future.

File: ttnn/cpp/ttnn/graph/graph_trace_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                    }
+                }
+                --i;  // adjust for loop increment
+            }
+            current_op.push_back(v["params"]["name"]);
+        } else if (v["name"] == "circular_buffer_allocate") {
+            total_cb += stoi(v["params"]["size"].get<std::string>());
+        } else if (v["name"] == "circular_buffer_deallocate_all") {
+            total_cb = 0;
+        } else if (v["name"] == "buffer_allocate" && v["params"]["type"] == "L1") {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/graph/graph_trace_utils.cpp:76]: what if type is DRAM?

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/graph/graph_trace_utils.cpp:76]: peak usage counting for L1 only, I will rename method to be clear

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/graph/graph_trace_utils.cpp:76]: thats what matters the most at the moment

Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/graph/graph_trace_utils.cpp:76]: got it, but I'd count for both. Less questions from the code users.

File: ttnn/cpp/ttnn/graph/graph_processor.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/graph/graph_processor.cpp:1]: this file is relocated, no changes

File: ttnn/cpp/ttnn/graph/graph_pybind.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/graph/graph_pybind.cpp:1]: file is relocated, not changes

File: ttnn/cpp/ttnn/graph/graph_trace_utils.cpp (hunk 3 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/graph/graph_trace_utils.cpp:76]: Leaving L1 for now

**Review by arakhmati — APPROVED** [review]

**Comment by ayerofieiev-tt** [comment]
We discussed and agreed with @dmakoviichuk-tt that we should add constants for node types

**Pull request merged**